OMCompiler/Compiler/Util/UnorderedMap.mo

```diff
@@ -0,0 +1,548 @@
+    Vector<K> keys;
+    Vector<V> values;
+    Hash hashFn;
+    KeyEq eqFn;
+  end UNORDERED_MAP;
+
+  function new<V>
+    "Creates a new map given a hash function, key equality function, and
+     optional desired bucket count. An appropriate bucket count is
+     Util.nextPrime(number of elements that will be added), but starting with a
+     low bucket count is also fine if the number of elements is unknown since
+     the map rehashes as needed."
+    input Hash hash;
+    input KeyEq keyEq;
+    input Integer bucketCount = 1;
+    output UnorderedMap<K, V> map;
+  protected
+    K kdummy = kdummy;
+    V vdummy = vdummy;
+  algorithm
+    map := UNORDERED_MAP(
+      Vector.newFill(bucketCount, {}),
+      Vector.new(kdummy),
+      Vector.new(vdummy),
+      hash,
+      keyEq
+    );
+  end new;
+
+  function copy
+    "Returns a copy of the map."
+    input UnorderedMap<K, V> map;
+    output UnorderedMap<K, V> outMap;
+  algorithm
+    outMap := UNORDERED_MAP(
+      Vector.copy(map.buckets),
+      Vector.copy(map.keys),
+      Vector.copy(map.values),
+      map.hashFn,
+      map.eqFn
+    );
+  end copy;
+
+  function deepCopy
+    "Returns a deep copy of the map using the given copy function for values."
+    input UnorderedMap<K, V> map;
+    input CopyFn fn;
+    output UnorderedMap<K, V> outMap;
+
+    partial function CopyFn
+      input output V value;
+    end CopyFn;
+  algorithm
+    outMap := UNORDERED_MAP(
+      Vector.copy(map.buckets),
+      Vector.copy(map.keys),
+      Vector.deepCopy(map.values, fn),
+      map.hashFn,
+      map.eqFn
+    );
+  end deepCopy;
+
+  function add
+    "Adds a key and associated value to the map, or updated the value if the key
+     already exists in the map. Might trigger a rehash."
+    input K key;
+    input V value;
+    input UnorderedMap<K, V> map;
+  protected
+    Integer index, hash;
+  algorithm
+    (index, hash) := find(key, map);
+
+    if index > 0 then
+      Vector.update(map.values, index, value);
+    else
+      addEntry(key, value, hash, map);
+    end if;
+  end add;
+
+  function addNew
+    "Adds a key and associated value to the map without checking if it already
+     exists. Faster than add since it doesn't need to check if the key exists,
+     but will lead to duplicate keys if it actually does exist in the map
+     already. Might trigger a rehash."
+    input K key;
+    input V value;
+    input UnorderedMap<K, V> map;
+  protected
+    Hash hashfn = map.hashFn;
+    Integer hash;
+  algorithm
+    hash := hashfn(key, Vector.size(map.buckets));
+    addEntry(key, value, hash, map);
+  end addNew;
+
+  function addUnique
+    "Adds a key and associated value to the map, but fails if the key already
+     exists. Might trigger a rehash."
+    input K key;
+    input V value;
+    input UnorderedMap<K, V> map;
+  protected
+    Integer index, hash;
+  algorithm
+    (index, hash) := find(key, map);
+    true := index > 0;
```

Comment:
I think this should fail when the key already exists, but now it does when it doesn't, right?
```suggestion
    false := index > 0;
```

Reply:
Yep, fixed in #6966.